### DISCUSSION TITLE
Conditionnaly use eager_default

### DISCUSSION THREAD

**Discussion by Smixi** [sqlalchemy/sqlalchemy#12142] [Usage Questions] [answered]
Hi !

I'm using using sqlalchemy with a declarative mapping with a postgres/asyncpg environment.

I have a parent class  defined like this:

```python
class Parent(DeclarativeBase):
    modified_at: Mapped[datetime] = mapped_column(
            DateTime(timezone=True),
            server_default=func.now(),
            onupdate=func.now(),
        )
    __mapper_args__: ClassVar = {"eager_defaults": True}
```

Some CRUD endpoints needs to have the modified_at fields fresh, while some others don't need it at all (even for processing).

My issue is that I have one endpoint modifying multiples items, which when flushed/commited together emit UPDATE ... RETURNING statement for each objects. Because I don't need the newly "modified_at" server side generated field later on, I would like to disable eager_default, specifically for this session so it can 

Is there any way to do that ?

Using False or "auto" is perfect for this usecase, but then some endpoints that needs the fields to be refreshed will fail, so I can't declare it globally.

Thanks in advance :).

**Comment by zzzeek** [comment]
there's no option for that level of granularity and if your DB supports RETURNING there's really no downside to just leaving it on

**Reply by Smixi** [reply]
In my case, I need to update N objects.

Lets say my object have a "value" field, and inherit my Parent class above (which thus have the modified_at field).
For instance:

```
for item in items:
    item.value = compute_new_value(item)
session.flush()
```

In this case, this will emit as much UPDATE ... RETURNING query as there is entry in the items list.
Using eager_default=False here leads to only one query, because no RETURNING is needed.
Because I know in advance that the modified_at field will not be used afterward, I don't care if it loaded or not (at least in this code section). So sometimes I would like to use False or True depending on wether the object needs the field or not after an update.

Maybe my use case is not best practice and I should use the ORM through session.execute ?

**Reply by zzzeek** [reply, marked as answer]
if you are trying to optimize specific operations, go all the way and use the facilities to render UPDATE statements in an ORM context, in this case https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#orm-bulk-update-by-primary-key 

demo

```py
from __future__ import annotations

from datetime import datetime

from sqlalchemy import annotations
from sqlalchemy import create_engine
from sqlalchemy import DateTime
from sqlalchemy import func
from sqlalchemy import update
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import Session


class Base(DeclarativeBase):
    pass


class Parent(Base):
    __tablename__ = 'parent'
    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[str]

    modified_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
    )
    __mapper_args__ = {"eager_defaults": True}


e = create_engine("postgresql://scott:tiger@localhost/test", echo=True)
Base.metadata.drop_all(e)
Base.metadata.create_all(e)

s = Session(e)

parents = [Parent(data=f"p{i}") for i in range(10)]

s.add_all(parents)

s.flush()

for p in parents:
    p.data += "new data 1"
# lots of UPDATE
s.flush()

# just one UPDATE
s.execute(
    update(Parent),
    [{
        "id": p.id, "data": p.data + " new data 2"
    } for p in parents]
)

s.commit()
```
